lib/ex_doc/refs.ex

```diff
@@ -132,8 +135,28 @@ defmodule ExDoc.Refs do
     end
   end
 
-  defp visibility(:hidden), do: :hidden
-  defp visibility(_), do: :public
+  defguardp is_kind(kind) when kind in [:module, :callback, :type, :function]
+  defguardp no_docs?(doc) when doc == :none or (is_map(doc) and map_size(doc) == 0)
+
+  defp visibility(:module, _name, doc) when no_docs?(doc), do: :public
+  defp visibility(:callback, _name, doc) when no_docs?(doc), do: :public
+  defp visibility(:type, _name, doc) when no_docs?(doc), do: :public
+
+  defp visibility(:function, name, doc) when no_docs?(doc) do
+    if hd(Atom.to_charlist(name)) == ?_ do
```

Comment:
This could be abstracted into a function, as it is also used in the Retriever module. But the question is where to place it. For now we just have duplicated code.